store/tikv/client_batch.go

```diff
@@ -193,26 +199,59 @@ func (a *batchConn) isIdle() bool {
 	return atomic.LoadUint32(&a.idle) != 0
 }
 
+func (a *batchConn) callbackResponse(resp *tikvpb.BatchCommandsResponse) {
+	if resp == nil {
+		a.failPendingRequests(context.Canceled)
+		return
+	}
+	a.pendingRequest -= len(resp.RequestIds)
+	for idx, id := range resp.RequestIds {
+		entry, ok := a.batched[id]
+		if !ok {
+			// this maybe caused by batchCommandsClient#send meets ambiguous error that request has be sent to TiKV but still report a error.
+			// then TiKV will send response back though stream and reach here.
+			logutil.BgLogger().Warn("batchRecvLoop receives outdated response", zap.Uint64("requestID", id))
+			a.pendingRequest += 1
+			continue
+		}
+		if atomic.LoadInt32(&entry.canceled) == 0 {
+			// Put the response only if the request is not canceled.
+			entry.res <- resp.Responses[idx]
+		}
+		delete(a.batched, id)
+		if trace.IsEnabled() {
+			trace.Log(entry.ctx, "rpc", "received")
+		}
+		logutil.Eventf(entry.ctx, "receive %T response with other %d batched requests", resp.Responses[idx].GetCmd(), len(resp.Responses))
+	}
+}
+
 // fetchAllPendingRequests fetches all pending requests from the channel.
 func (a *batchConn) fetchAllPendingRequests(
 	maxBatchSize int,
+	maxWaitTime time.Duration,
 ) time.Time {
 	// Block on the first element.
 	var headEntry *batchCommandsEntry
-	select {
-	case headEntry = <-a.batchCommandsCh:
-		if !a.idleDetect.Stop() {
-			<-a.idleDetect.C
-		}
-		a.idleDetect.Reset(idleTimeout)
-	case <-a.idleDetect.C:
-		a.idleDetect.Reset(idleTimeout)
-		atomic.AddUint32(&a.idle, 1)
-		atomic.CompareAndSwapUint32(a.idleNotify, 0, 1)
-		// This batchConn to be recycled
-		return time.Now()
-	case <-a.closed:
-		return time.Now()
+	for headEntry == nil {
+		select {
+		case headEntry = <-a.batchCommandsCh:
+			if !a.idleDetect.Stop() {
+				<-a.idleDetect.C
+			}
+			a.idleDetect.Reset(idleTimeout)
+			break
+		case <-a.idleDetect.C:
+			a.idleDetect.Reset(idleTimeout)
+			atomic.AddUint32(&a.idle, 1)
+			atomic.CompareAndSwapUint32(a.idleNotify, 0, 1)
+			// This batchConn to be recycled
+			return time.Now()
+		case <-a.closed:
```

Comment:
do we need to exhaust the `a.resps` when `a.closed` channel is closed~?

for example, some situation like 

1. `a.resps` is full with `len(a.resps) == maxBatchSize*4`, this make all `recvLoop` goroutine blocked when `a.resps <- resp`
2. tidb graceful shutdown

the question is:

- requests in `a.resps` have no chance to return to invoker until some long timeout
- `recvLoop` goroutine maybe block forever, because it block on send channel and have no chance to check `isStop()`...maybe change block send `a.resps <- resp` to `select case a.resps <- ..` also can solve it